addons/crm/static/src/js/forecast/forecast_service.js

```diff
@@ -0,0 +1,230 @@
+        this.minGranularity = fieldType === 'datetime' ? "second" : "day";
+        this.set_min_groups(min_groups); 
+
+        this._computeStart();
+        this._computeEnd();
+    }
+    /**
+     * @see GRANULARITY_TABLE
+     */
+    _get(setting) {
+        return GRANULARITY_TABLE[this.granularity][setting];
+    }
+    /**
+     * Compute the moment for the start of the period containing "now"
+     */
+    _computeStart() {
+        this.start = this._get("startOf")(moment());
+    }
+    /**
+     * Compute the moment for the end of the forecast period. This bound is exclusive.
+     * The forecast period is the number of [granularity] from [start] to the end of the [cycle] reached after
+     * adding [min_groups] 
+     * i.e. we are in october 2020 :
+     *      [start] = 2020-10-01
+     *      [granularity] = 'month',
+     *      [cycle] = 12
+     *      [min_groups] = 4,
+     *      => forecastPeriod = 15 months (until end of december 2021)
+     */
+    _computeEnd() {
+        var cycle = this._get("cycle");
+        var cyclePos = this._get("cyclePos")(this.start);
+        /**
+         * forecastPeriod formula explanation :
+         * We want to know how many steps need to be taken from the current position until the end of the cycle
+         * reached after guaranteeing min_groups positions. Let's call this cycle (C).
+         * 
+         * (1) compute the steps needed to reach the last position of the current cycle, from the current position:
+         *     {cycle - cyclePos}
+         * 
+         * (2) ignore {min_groups - 1} steps from the position reached in (1). Now, the current position is somewhere
+         *     in (C). One step from min_groups is reserved to reach the first position after (C), hence {-1}
+         * 
+         * (3) compute the additional steps needed to reach the last position of (C), from the position reached in (2):
+         *     {cycle - (min_groups - 1) % cycle}
+         * 
+         * (4) combine (1) and (3), the sum should not be greater than a full cycle (-> truncate):
+         *     {(2 * cycle - (min_groups - 1) % cycle - cyclePos) % cycle}
+         * 
+         * (4) add min_groups!
+         */
+        var forecastPeriod = (2 * cycle - (this.min_groups - 1) % cycle - cyclePos) % cycle + this.min_groups;
+        this.end = moment(this.start).add(forecastPeriod, `${this.granularity}s`);
+    }
+    /**
+     * Adds one "granularity" period to [this.end], to expand the current forecast period
+     */
+    expand() {
+        this.end.add(1, `${this.granularity}s`);
+    }
+    /**
+     * @param {Array[]} domain
+     * @param {boolean} limited whether [this.end] moment must be used as a domain constraint to limit
+     *                          read_group results [limited === true]
+     *                          or not [limited === false]
+     * @returns {Array[]} modified domain
+     */
+    getDomain(domain, limited=undefined) {
```

Comment:
This is because the `fill_temporal` context key in it's `dictionary` form will only return contiguous groups between the chosen bounds, but will not prevent groups outside those bounds from being returned by the read_group. To do this, we have to add the same constraints to the domain, so that only the contiguous groups are returned. 

But as mentioned in the comment above, there is one case when we still want those "outsider groups", that is when we want to clean "forgotten old opportunities" by disabling the filter from the `forecastModelExtension`

Reply:
I will use `force_end_limit` instead of `limited`